be/src/formats/parquet/file_writer.cpp

```diff
@@ -177,136 +191,645 @@ void ParquetBuildHelper::build_compression_type(::parquet::WriterProperties::Bui
+        nulls = null_column->get_data().data();
+    }
+
+    const auto map_column = down_cast<MapColumn*>(ColumnHelper::get_data_column(col.get()));
+    const auto keys = map_column->keys_column();
+    const auto values = map_column->values_column();
+    const auto offsets = map_column->offsets_column()->get_data();
+
+    Context key_ctx(ctx._max_rep_level + 1, ctx.size() + keys->size());
+    Context value_ctx(ctx._max_rep_level + 1, ctx.size() + values->size());
+
+    int subcol_size = 0;
+    for (auto i = 0; i < ctx.size(); i++) {
+        auto [idx, def_level, rep_level] = ctx.get(i);
+        if (idx == Context::kNULL || nulls[idx]) {
+            key_ctx.append(Context::kNULL, def_level, rep_level);
+            value_ctx.append(Context::kNULL, def_level, rep_level);
+            continue;
         }
-        case TYPE_CHAR:
-        case TYPE_VARCHAR: {
-            DISPATCH_PARQUET_STRING_WRITER()
-            break;
+        auto map_size = offsets[idx + 1] - offsets[idx];
+        if (map_size == 0) {
+            key_ctx.append(Context::kNULL, def_level + outer_node->is_optional(), rep_level);
+            key_ctx.append(Context::kNULL, def_level + outer_node->is_optional(), rep_level);
+            continue;
         }
-        default: {
-            //TODO: support other types
-            return Status::InvalidArgument("Unsupported type");
+        key_ctx.append(subcol_size++, def_level + outer_node->is_optional() + key_node->is_optional(), rep_level);
+        key_ctx.append(subcol_size++, def_level + outer_node->is_optional() + value_node->is_optional(), rep_level);
+        for (auto offset = 1; offset < map_size; offset++) {
+            key_ctx.append(subcol_size++, def_level + outer_node->is_optional() + key_node->is_optional(),
+                           key_ctx._max_rep_level);
+            key_ctx.append(subcol_size++, def_level + outer_node->is_optional() + value_node->is_optional(),
+                           value_ctx._max_rep_level);
         }
+    }
+
+    auto ret = _add_column_chunk(key_ctx, type_desc.children[0], key_node, keys);
+    if (!ret.ok()) {
+        return ret;
+    }
+    return _add_column_chunk(value_ctx, type_desc.children[1], value_node, values);
+}
+
+Status FileWriterBase::_add_varchar_column_chunk(const Context& ctx, const TypeDescriptor& type_desc,
+                                                 const ::parquet::schema::NodePtr& node, const ColumnPtr& col) {
+    unsigned char* nulls = nullptr;
+    if (col->is_nullable()) {
+        const auto null_column = down_cast<NullableColumn*>(col.get())->null_column();
+        nulls = null_column->get_data().data();
+    }
+
+    const auto data_column = ColumnHelper::get_data_column(col.get());
+    auto raw_col = down_cast<const RunTimeColumnType<TYPE_VARCHAR>*>(data_column);
+    auto vo = raw_col->get_offset();
+    auto vb = raw_col->get_bytes();
+
+    _generate_rg_writer();
```

Comment:
generate row group writer before _add_column_chunk()

Reply:
Yep. This line can be removed.